File: apps/mobile/ios/SpeakeasyBridges/Signal/SignalProtocolModule.swift
//
//  SignalProtocolModule.swift
//  Speakeasy
//
//  Phase 5b iOS — RN bridge for the 1:1 Signal Protocol path. Mirrors
//  apps/mobile/android/.../signal/SignalProtocolModule.kt
//
//  # libsignal-client iOS API NOTE
//
//  The Swift API uses free functions (`processPreKeyBundle`,
//  `signalEncrypt`, `signalDecryptPreKey`, `signalDecrypt`) instead of
//  the `SessionBuilder` / `SessionCipher` classes the Java/Kotlin API
//  exposes. Each takes the relevant store protocols as parameters.
//  Anything marked `// libsignal:` is a best-guess based on the
//  reference docs; the exact signatures may need a one-pass fix once
//  Mac access lands.
//
//  # Wire format (matches Android exactly)
//
//  encrypt: returns base64( typeByte || libsignalSerializedMessage )
//    typeByte 0x03 = PreKeySignalMessage (first message after handshake)
//    typeByte 0x02 = SignalMessage       (subsequent messages)
//
//  decrypt: takes base64( typeByte || body ), dispatches based on the byte.
//

import Foundation
import LibSignalClient

@objc(SignalProtocolModule)
class SignalProtocolModule: NSObject {

    @objc static func requiresMainQueueSetup() -> Bool { return false }

    private let deviceId: UInt32 = 1
    private let context = NullContext()

    // MARK: - generateIdentityKey

    @objc(generateIdentityKey:rejecter:)
    func generateIdentityKey(_ resolve: @escaping RCTPromiseResolveBlock,
                             rejecter reject: @escaping RCTPromiseRejectBlock) {
        do {
            // Don't clobber a persisted identity — restore + return its
            // public key instead of minting a fresh one.
            if SpeakeasySignalStore.initializeFromDb() {
                let store = try SpeakeasySignalStore.require()
                let ikp = try store.identityKeyPair(context: context)
                resolve(Data(ikp.publicKey.serialize()).base64EncodedString())
                return
            }
            let ikp = IdentityKeyPair.generate()
            // Match Android's range: 1..16380 (libsignal's reserved space).
            let regId = UInt32.random(in: 1...16380)
            try SpeakeasySignalStore.initialize(identityKeyPair: ikp, registrationId: regId)
            resolve(Data(ikp.publicKey.serialize()).base64EncodedString())
        } catch {
            reject("identity_key_failed", error.localizedDescription, error)
        }
    }

    // MARK: - generatePreKeyBundle

    @objc(generatePreKeyBundle:signedPreKeyId:oneTimePreKeyCount:resolver:rejecter:)
    func generatePreKeyBundle(_ registrationIdNum: NSNumber,
                              signedPreKeyId: NSNumber,
                              oneTimePreKeyCount: NSNumber,
                              resolver resolve: @escaping RCTPromiseResolveBlock,
                              rejecter reject: @escaping RCTPromiseRejectBlock) {
        do {
            ensureRestored()
            let store = try SpeakeasySignalStore.require()
            let ikp = try store.identityKeyPair(context: context)

            // Signed prekey — long-lived, server-stored, signed by identity.
            let signedKeyId = UInt32(truncating: signedPreKeyId)
            let signedPreKeyPair = PrivateKey.generate()
            let signedPreKeyPub = signedPreKeyPair.publicKey
            let signedPreKeySig = ikp.privateKey.generateSignature(
                message: signedPreKeyPub.serialize()
            )
            let signedRecord = try SignedPreKeyRecord(
                id: signedKeyId,
                timestamp: UInt64(Date().timeIntervalSince1970 * 1000),
                privateKey: signedPreKeyPair,
                signature: Data(signedPreKeySig)
            )
            try store.storeSignedPreKey(signedRecord, id: signedKeyId, context: context)

            // One-time prekeys.
            var preKeysOut: [[String: Any]] = []
            let count = Int(truncating: oneTimePreKeyCount)
            for i in 0..<count {
                let pkId = UInt32(i + 1)
                let pkPair = PrivateKey.generate()
                let pkRec = try PreKeyRecord(id: pkId, privateKey: pkPair)
                try store.storePreKey(pkRec, id: pkId, context: context)
                preKeysOut.append([
                    "id": Int(pkId),
                    "key": Data(pkPair.publicKey.serialize()).base64EncodedString()
                ])
            }

            resolve([
                "registrationId":     Int(truncating: registrationIdNum),
                "signedPreKeyId":     Int(signedKeyId),
                "signedPreKey":       Data(signedPreKeyPub.serialize()).base64EncodedString(),
                "signedPreKeySig":    Data(signedPreKeySig).base64EncodedString(),
                "preKeys":            preKeysOut,
                "identityPublicKey":  Data(ikp.publicKey.serialize()).base64EncodedString()
            ])
        } catch {
            reject("prekey_bundle_failed", error.localizedDescription, error)
        }
    }

    // MARK: - initiateSession

    @objc(initiateSession:peerBundle:resolver:rejecter:)
    func initiateSession(_ peerUserId: NSString,
                         peerBundle: NSDictionary,
                         resolver resolve: @escaping RCTPromiseResolveBlock,
                         rejecter reject: @escaping RCTPromiseRejectBlock) {
        do {
            ensureRestored()
            let store = try SpeakeasySignalStore.require()
            let regId = UInt32(truncating: peerBundle["registrationId"] as? NSNumber ?? 0)
            let signedKeyId = UInt32(truncating: peerBundle["signedPreKeyId"] as? NSNumber ?? 0)
            guard
                let signedKeyB64 = peerBundle["signedPreKey"] as? String,
                let signedSigB64 = peerBundle["signedPreKeySig"] as? String,
                let identityB64 = peerBundle["identityPublicKey"] as? String,
                let preKeys = peerBundle["preKeys"] as? [[String: Any]],
                let firstPK = preKeys.first,
                let firstPKKeyB64 = firstPK["key"] as? String,
                let firstPKIdNum = firstPK["id"] as? NSNumber
            else {
                reject("session_init_failed", "malformed peer bundle", nil)
                return
            }
            let identityKey = try IdentityKey(bytes: Array(b64(identityB64)))
            let signedPreKey = try PublicKey(Array(b64(signedKeyB64)))
            let oneTimePreKey = try PublicKey(Array(b64(firstPKKeyB64)))

            let bundle = try PreKeyBundle(
                registrationId: regId,
                deviceId: deviceId,
                prekeyId: UInt32(truncating: firstPKIdNum),
                prekey: oneTimePreKey,
                signedPrekeyId: signedKeyId,
                signedPrekey: signedPreKey,
                signedPrekeySignature: Array(b64(signedSigB64)),
                identity: identityKey
            )

            let peerAddr = try ProtocolAddress(name: peerUserId as String, deviceId: deviceId)
            // libsignal: free function. Persists session into the store.
            try processPreKeyBundle(
                bundle,
                for: peerAddr,
                sessionStore: store,
                identityStore: store,
                context: context
            )
            resolve(NSNull())
        } catch SignalError.untrustedIdentity(_) {
            reject("untrusted_identity", "peer's identity key changed", nil)
        } catch {
            reject("session_init_failed", error.localizedDescription, error)
        }
    }

    // MARK: - encrypt / decrypt

    @objc(encrypt:plaintext:resolver:rejecter:)
    func encrypt(_ peerUserId: NSString,
                 plaintext plaintextB64: NSString,
                 resolver resolve: @escaping RCTPromiseResolveBlock,
                 rejecter reject: @escaping RCTPromiseRejectBlock) {
        do {
            ensureRestored()
            let store = try SpeakeasySignalStore.require()
            let peerAddr = try ProtocolAddress(name: peerUserId as String, deviceId: deviceId)
            let plaintext = b64(plaintextB64 as String)

            // libsignal: free-function encrypt. Returns CiphertextMessage
            // which has `.messageType` (PreKey or Whisper) + `.serialize()`.
            let msg = try signalEncrypt(
                message: plaintext,
                for: peerAddr,
                sessionStore: store,
                identityStore: store,
                context: context
            )
            let typeByte: UInt8 = (msg.messageType == .preKey) ? 0x03 : 0x02
            var out = Data()
            out.append(typeByte)
            out.append(Data(msg.serialize()))
            resolve(out.base64EncodedString())
        } catch {
            reject("encrypt_failed", error.localizedDescription, error)
        }
    }

    @objc(decrypt:ciphertext:resolver:rejecter:)
    func decrypt(_ peerUserId: NSString,
                 ciphertext ciphertextB64: NSString,
                 resolver resolve: @escaping RCTPromiseResolveBlock,
                 rejecter reject: @escaping RCTPromiseRejectBlock) {
        do {
            ensureRestored()
            let store = try SpeakeasySignalStore.require()
            let peerAddr = try ProtocolAddress(name: peerUserId as String, deviceId: deviceId)
            let raw = b64(ciphertextB64 as String)
            guard raw.count > 1 else {
                reject("decrypt_failed", "empty ciphertext", nil)
                return
            }
            let typeByte = raw[0]
            let body = Array(raw.suffix(from: 1))

            guard typeByte == 0x03 || typeByte == 0x02 else {
                reject("decrypt_failed", "unknown ciphertext type byte 0x\(String(typeByte, radix: 16))", nil)
                return
            }

            // Idempotent decrypt: a re-presented ciphertext (dropped WS ack →
            // server redelivery on reconnect, or — once it ships — the push
            // NSE decrypting the same message) returns the cached plaintext
            // instead of re-running the single-use Double Ratchet, which would
            // throw duplicatedMessage and surface a spurious "[decrypt failed]"
            // bubble. The ratchet still runs exactly once on a true first
            // decrypt; its errors propagate unchanged. Mirrors Android's
            // DecryptCache (signal/DecryptCache.kt).
            let plaintext = try DecryptCache.decryptCached(ciphertext: Array(raw)) {
                if typeByte == 0x03 {
                    let msg = try PreKeySignalMessage(bytes: body)
                    // libsignal: free-function decrypt for PreKey message.
                    return try signalDecryptPreKey(
                        message: msg,
                        from: peerAddr,
                        sessionStore: store,
                        identityStore: store,
                        preKeyStore: store,
                        signedPreKeyStore: store,
                        kyberPreKeyStore: store,
                        context: context
                    )
                } else {
                    let msg = try SignalMessage(bytes: body)
                    return try signalDecrypt(
                        message: msg,
                        from: peerAddr,
                        sessionStore: store,
                        identityStore: store,
                        context: context
                    )
                }
            }
            resolve(Data(plaintext).base64EncodedString())
        } catch SignalError.untrustedIdentity(_) {
            reject("untrusted_identity", "peer's identity key changed", nil)
        } catch {
            reject("decrypt_failed", error.localizedDescription, error)
        }
    }

    // MARK: - hasSession

    /// Is there a persisted Signal session row for `peerUserId`
    /// (device 1)? Lets the JS `ensureSessionWithPeer` skip the
    /// destructive `initiateSession` re-init when a session already
    /// exists on disk from a prior process. Without it, a cold-start-
    /// send-first burns a fresh OTPK and emits a PreKeySignalMessage
    /// the peer's already-advanced libsignal rejects ("invalid PreKey
    /// message: decryption failed").
    ///
    /// iOS has no `containsSession` on the SessionStore protocol —
    /// `loadSession` returns nil when no row exists (see
    /// SqlCipherSignalProtocolStore.loadSession).
    @objc(hasSession:resolver:rejecter:)
    func hasSession(_ peerUserId: NSString,
                    resolver resolve: @escaping RCTPromiseResolveBlock,
                    rejecter reject: @escaping RCTPromiseRejectBlock) {
        do {
            ensureRestored()
            let store = try SpeakeasySignalStore.require()
            let peerAddr = try ProtocolAddress(name: peerUserId as String, deviceId: deviceId)
            let session = try store.loadSession(for: peerAddr, context: context)
            resolve(session != nil)
        } catch {
            reject("has_session_failed", error.localizedDescription, error)
        }
    }

    // MARK: - Peer / store reset

    /// Forget a peer's saved identity + session so the next
    /// initiateSession re-TOFUs their rotated identity. Backs the
    /// "[couldn't decrypt] → reset" recovery path (ChatScreen). Mirrors
    /// Android SignalProtocolModule.resetPeer.
    @objc(resetPeer:resolver:rejecter:)
    func resetPeer(_ peerUserId: NSString,
                   resolver resolve: @escaping RCTPromiseResolveBlock,
                   rejecter reject: @escaping RCTPromiseRejectBlock) {
        do {
            ensureRestored()
            let store = try SpeakeasySignalStore.require()
            try store.clearPeerIdentity(peerUserId as String)
            resolve(nil)
        } catch {
            reject("reset_peer_failed", error.localizedDescription, error)
        }
    }

    /// Permanently wipe the entire Signal store — drop the in-memory
    /// handle and delete the encrypted SQLCipher DB. Backs account
    /// deletion so a re-enrollment starts from an empty store. Mirrors
    /// Android SignalProtocolModule.wipeStore.
    @objc(wipeStore:rejecter:)
    func wipeStore(_ resolve: @escaping RCTPromiseResolveBlock,
                   rejecter reject: @escaping RCTPromiseRejectBlock) {
        SpeakeasySignalStore.reset()
        SpeakeasyDb.shared.wipe()
        resolve(nil)
    }

    // MARK: - Helpers

    /// Lazy on-disk restore, mirrors Android's ensureRestored().
    private func ensureRestored() {
        if !SpeakeasySignalStore.isInitialized {
            _ = SpeakeasySignalStore.initializeFromDb()
        }
    }

    private func b64(_ s: String) -> Data {
        return Data(base64Encoded: s) ?? Data()
    }
}

File: apps/mobile/android/app/src/main/java/xyz/speakeasyapp/app/signal/SignalProtocolModule.kt
package xyz.speakeasyapp.app.signal

import android.util.Base64
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import xyz.speakeasyapp.app.db.SpeakeasyDb
import org.signal.libsignal.protocol.IdentityKey
import org.signal.libsignal.protocol.IdentityKeyPair
import org.signal.libsignal.protocol.SessionBuilder
import org.signal.libsignal.protocol.SessionCipher
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.signal.libsignal.protocol.UntrustedIdentityException
import org.signal.libsignal.protocol.ecc.Curve
import org.signal.libsignal.protocol.ecc.ECPublicKey
import org.signal.libsignal.protocol.message.PreKeySignalMessage
import org.signal.libsignal.protocol.message.SignalMessage
import org.signal.libsignal.protocol.state.PreKeyBundle
import org.signal.libsignal.protocol.state.PreKeyRecord
import org.signal.libsignal.protocol.state.SignedPreKeyRecord

/**
 * RN bridge for the Signal Protocol (`org.signal:libsignal-android`).
 *
 * Phase 5b: covers the 1:1 path — generateIdentityKey, generatePreKeyBundle,
 * initiateSession, encrypt, decrypt. Sender Keys for groups
 * (`encryptForGroup`/`decryptFromGroupMember`) are deferred to the next
 * sweep — see spec §11 Phase 5b carry-over.
 *
 * Storage: Phase 5c swaps the in-memory `InMemorySignalProtocolStore` for
 * a SQLCipher-backed [SqlCipherSignalProtocolStore] keyed by an Android
 * Keystore-wrapped passphrase (see `SpeakeasyDb` for the deviation note vs
 * the spec's "Vouchflow device key" wording). The first call into any
 * bridge method auto-restores from the encrypted DB if an identity row
 * already exists, so cold starts no longer require re-enrollment.
 *
 * Wire format:
 *   - All bytes cross the bridge as base64 strings (RN bridge limitation).
 *   - encrypt/decrypt prefix the ciphertext with a 1-byte type marker so
 *     the recipient knows which libsignal message class to decode:
 *       0x03 = PreKeySignalMessage  (first message after initiateSession)
 *       0x02 = SignalMessage         (subsequent messages)
 *
 * Errors: rejected with code = a stable string the JS layer maps to
 * `SignalClientErrorReason`. Every code is documented at the JS interface.
 */
class SignalProtocolModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

  override fun getName(): String = "SignalProtocol"

  // Default device id; multi-device fan-out is server-side only in Phase 4.
  private val deviceId: Int = 1

  private fun b64(bytes: ByteArray): String =
      Base64.encodeToString(bytes, Base64.NO_WRAP)

  private fun unb64(s: String): ByteArray = Base64.decode(s, Base64.NO_WRAP)

  /**
   * Lazily restore the persistent store from disk on first use. No-op if
   * the store is already initialized (post-restore or post-generate). Safe
   * to call from any bridge method.
   */
  private fun ensureRestored() {
    if (!SpeakeasySignalStore.isInitialized()) {
      SpeakeasySignalStore.initializeFromDb(reactContext)
    }
  }

  @ReactMethod
  fun generateIdentityKey(promise: Promise) {
    try {
      // Phase 5c: don't clobber an already-persisted identity. If one is on
      // disk, restore + return its public key instead of minting a fresh one.
      if (SpeakeasySignalStore.initializeFromDb(reactContext)) {
        val ikp = SpeakeasySignalStore.requireInitialized().identityKeyPair
        promise.resolve(b64(ikp.publicKey.serialize()))
        return
      }
      val identityKeyPair = IdentityKeyPair.generate()
      val registrationId = (1..16380).random()
      SpeakeasySignalStore.initialize(reactContext, identityKeyPair, registrationId)
      promise.resolve(b64(identityKeyPair.publicKey.serialize()))
    } catch (e: Throwable) {
      promise.reject("identity_key_failed", e.message, e)
    }
  }

  @ReactMethod
  fun generatePreKeyBundle(
      registrationId: Int,
      signedPreKeyId: Int,
      oneTimePreKeyCount: Int,
      promise: Promise,
  ) {
    try {
      ensureRestored()
      val store = SpeakeasySignalStore.requireInitialized()
      val identityKeyPair = store.identityKeyPair

      // Signed prekey — long-lived, server-stored, signed by identity.
      val signedPreKeyPair = Curve.generateKeyPair()
      val signedPreKeySig =
          Curve.calculateSignature(
              identityKeyPair.privateKey, signedPreKeyPair.publicKey.serialize())
      val signedRecord =
          SignedPreKeyRecord(
              signedPreKeyId,
              System.currentTimeMillis(),
              signedPreKeyPair,
              signedPreKeySig)
      store.storeSignedPreKey(signedPreKeyId, signedRecord)

      // One-time prekeys — bucket the server hands out to peers establishing
      // new sessions. Each gets consumed once (server side).
      val preKeysArr = Arguments.createArray()
      for (i in 0 until oneTimePreKeyCount) {
        val pkId = i + 1
        val pkPair = Curve.generateKeyPair()
        store.storePreKey(pkId, PreKeyRecord(pkId, pkPair))
        preKeysArr.pushMap(
            Arguments.createMap().apply {
              putInt("id", pkId)
              putString("key", b64(pkPair.publicKey.serialize()))
            })
      }

      promise.resolve(
          Arguments.createMap().apply {
            putInt("registrationId", registrationId)
            putInt("signedPreKeyId", signedPreKeyId)
            putString("signedPreKey", b64(signedPreKeyPair.publicKey.serialize()))
            putString("signedPreKeySig", b64(signedPreKeySig))
            putArray("preKeys", preKeysArr)
            // Identity public key — server stores per-user.
            putString("identityPublicKey", b64(identityKeyPair.publicKey.serialize()))
          })
    } catch (e: Throwable) {
      promise.reject("prekey_bundle_failed", e.message, e)
    }
  }

  @ReactMethod
  fun initiateSession(peerUserId: String, peerBundle: ReadableMap, promise: Promise) {
    try {
      ensureRestored()
      val store = SpeakeasySignalStore.requireInitialized()

      val regId = peerBundle.getInt("registrationId")
      val signedPreKeyId = peerBundle.getInt("signedPreKeyId")
      val signedPreKeyBytes = unb64(peerBundle.getString("signedPreKey")!!)
      val signedPreKeySig = unb64(peerBundle.getString("signedPreKeySig")!!)
      val identityPublicKeyBytes = unb64(peerBundle.getString("identityPublicKey")!!)

      // Pick the first one-time prekey the server returned. (Server's
      // `/v1/prekeys/bundle` already consumed exactly one for us.)
      val preKeysArr = peerBundle.getArray("preKeys")
      if (preKeysArr == null || preKeysArr.size() == 0) {
        promise.reject("no_prekey", "peer bundle missing one-time prekey")
        return
      }
      val firstPreKey = preKeysArr.getMap(0)
      val preKeyId = firstPreKey.getInt("id")
      val preKeyBytes = unb64(firstPreKey.getString("key")!!)

      val identityKey = IdentityKey(identityPublicKeyBytes, 0)
      val signedPreKeyPub = Curve.decodePoint(signedPreKeyBytes, 0)
      val preKeyPub: ECPublicKey = Curve.decodePoint(preKeyBytes, 0)

      val bundle =
          PreKeyBundle(
              regId,
              deviceId,
              preKeyId,
              preKeyPub,
              signedPreKeyId,
              signedPreKeyPub,
              signedPreKeySig,
              identityKey,
          )

      val peerAddress = SignalProtocolAddress(peerUserId, deviceId)
      val builder = SessionBuilder(store, peerAddress)
      builder.process(bundle)
      promise.resolve(null)
    } catch (e: UntrustedIdentityException) {
      promise.reject("untrusted_identity", e.message, e)
    } catch (e: Throwable) {
      promise.reject("session_init_failed", e.message, e)
    }
  }

  @ReactMethod
  fun encrypt(peerUserId: String, plaintextB64: String, promise: Promise) {
    try {
      ensureRestored()
      val store = SpeakeasySignalStore.requireInitialized()
      val plaintext = unb64(plaintextB64)
      val cipher = SessionCipher(store, SignalProtocolAddress(peerUserId, deviceId))
      val msg = cipher.encrypt(plaintext)
      // Prefix with type byte so the recipient can dispatch decrypt cleanly.
      val typeByte = msg.type.toByte()
      val out = ByteArray(1 + msg.serialize().size)
      out[0] = typeByte
      System.arraycopy(msg.serialize(), 0, out, 1, msg.serialize().size)
      promise.resolve(b64(out))
    } catch (e: Throwable) {
      promise.reject("encrypt_failed", e.message, e)
    }
  }

  @ReactMethod
  fun decrypt(peerUserId: String, ciphertextB64: String, promise: Promise) {
    try {
      ensureRestored()
      val store = SpeakeasySignalStore.requireInitialized()
      val raw = unb64(ciphertextB64)
      if (raw.isEmpty()) {
        promise.reject("decrypt_failed", "empty ciphertext")
        return
      }
      // Idempotent decrypt: the headless push handler and the in-app WS
      // path both decrypt the same ciphertext; the ratchet may advance
      // only once. See DecryptCache.
      val plaintext =
          DecryptCache.decryptCached(SpeakeasyDb.open(reactContext), raw) {
            val typeByte = raw[0].toInt() and 0xFF
            val body = raw.copyOfRange(1, raw.size)
            val cipher = SessionCipher(store, SignalProtocolAddress(peerUserId, deviceId))
            when (typeByte) {
              3 -> cipher.decrypt(PreKeySignalMessage(body))
              2 -> cipher.decrypt(SignalMessage(body))
              else ->
                  throw IllegalArgumentException("unknown ciphertext type byte $typeByte")
            }
          }
      promise.resolve(b64(plaintext))
    } catch (e: UntrustedIdentityException) {
      promise.reject("untrusted_identity", e.message, e)
    } catch (e: Throwable) {
      promise.reject("decrypt_failed", e.message, e)
    }
  }

  /**
   * Clear all stored state about a peer — their identity key + every
   * session record. Used to recover from `UntrustedIdentityException`
   * when the user opts in to trust a peer's freshly-rotated identity
   * (typical after the peer reinstalls / re-enrolls and gets a new
   * Signal identity key).
   *
   * After this returns, the next `initiateSession(peerUserId, …)`
   * fetches the peer's current PreKey bundle and saves the new identity
   * via TOFU — exactly as if we had never communicated with them.
   *
   * The caller is responsible for also dropping any in-process session
   * cache (see `crypto/session.ts:initiatedPeers`).
   */
  @ReactMethod
  fun resetPeer(peerUserId: String, promise: Promise) {
    try {
      ensureRestored()
      val store = SpeakeasySignalStore.requireInitialized()
      // SpeakeasySignalStore returns the SignalProtocolStore interface;
      // clearPeerIdentity is a SqlCipher-specific helper. There's only
      // one impl in production (the InMemoryStore is test-only).
      (store as SqlCipherSignalProtocolStore).clearPeerIdentity(peerUserId)
      promise.resolve(null)
    } catch (e: Throwable) {
      promise.reject("reset_peer_failed", e.message, e)
    }
  }
}
